### DISCUSSION TITLE
"ejbca.sh ca" commands throw exception after upgrade to EJBCA CE 8.0

### DISCUSSION THREAD

**Discussion by ecarlseen** [Keyfactor/ejbca-ce#421] [Q&A]
After updating from CE 7.11.0 to CE 8.0.20230531 (on Wildfly 26 / OpenJDK 11) I'm no longer able to manipulate the CAs using the CLI. Other functionality appears to be normal so far. I haven't yet run across any issues in the web UI.

For example, if I run the command :

> ejbca.sh ca listcas

I get the error below. I get a similar error with any other ca related commands that I've tried so far. This is part of a script that generates CRLs and uploads them to distribution servers as par of a cron job. Any suggestions appreciated.

> 

> Exception in thread "main" javax.ejb.EJBException: Failed to read response
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.apply(EJBClientChannel.java:1347)
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.getResult(EJBClientChannel.java:1356)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:620)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.protocol.remote.RemotingEJBClientInterceptor.handleInvocationResult(RemotingEJBClientInterceptor.java:57)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.TransactionPostDiscoveryInterceptor.handleInvocationResult(TransactionPostDiscoveryInterceptor.java:148)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.DiscoveryEJBClientInterceptor.handleInvocationResult(DiscoveryEJBClientInterceptor.java:130)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.NamingEJBClientInterceptor.handleInvocationResult(NamingEJBClientInterceptor.java:87)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.AuthenticationContextEJBClientInterceptor.call(AuthenticationContextEJBClientInterceptor.java:59)
> 	at org.jboss.ejb.client.AuthenticationContextEJBClientInterceptor.handleInvocationResult(AuthenticationContextEJBClientInterceptor.java:52)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.TransactionInterceptor.handleInvocationResult(TransactionInterceptor.java:212)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551)
> 	at org.jboss.ejb.client.EJBClientInvocationContext.awaitResponse(EJBClientInvocationContext.java:1003)
> 	at org.jboss.ejb.client.EJBInvocationHandler.invoke(EJBInvocationHandler.java:182)
> 	at org.jboss.ejb.client.EJBInvocationHandler.invoke(EJBInvocationHandler.java:116)
> 	at com.sun.proxy.$Proxy11.getCAInfo(Unknown Source)
> 	at org.ejbca.ui.cli.ca.CaListCAsCommand.execute(CaListCAsCommand.java:51)
> 	at org.ejbca.ui.cli.infrastructure.command.PasswordUsingCommandBase.execute(PasswordUsingCommandBase.java:201)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:300)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:311)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary.findAndExecuteCommandFromParameters(CommandLibrary.java:86)
> 	at org.ejbca.ui.cli.EjbcaEjbCli.main(EjbcaEjbCli.java:33)
> Caused by: java.lang.ClassNotFoundException: com.keyfactor.util.CertificateSerializableWrapper
> 	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
> 	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
> 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:527)
> 	at java.base/java.lang.Class.forName0(Native Method)
> 	at java.base/java.lang.Class.forName(Class.java:398)
> 	at org.jboss.marshalling.AbstractClassResolver.loadClass(AbstractClassResolver.java:129)
> 	at org.jboss.marshalling.AbstractClassResolver.resolveClass(AbstractClassResolver.java:110)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1048)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1381)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadCollectionObject(RiverUnmarshaller.java:202)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.readCollectionData(RiverUnmarshaller.java:876)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:752)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.readFields(RiverUnmarshaller.java:1879)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1793)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doInitSerializable(RiverUnmarshaller.java:1741)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1421)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298)
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:231)
> 	at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41)
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.apply(EJBClientChannel.java:1323)
> 	... 32 more
>

**Comment by ecarlseen** [comment]
UPDATE:

If I run a ca command that doesn't produce meaningful output I don't get an error. For example, if I run:

> ejbca.sh ca listexpired -d 10

I get this result, because I have no certificates expiring within the next ten days:

> Looking for certificates that expire before 'Mon Dec 04 21:50:52 PST 2023', but that are not expired now (Fri Nov 24 21:50:52 PST 2023).

However, if I increase this to the point where I will return results (set to 1000 days just for kicks):

> ejbca.sh ca listexpired -d 1000

I get this error:

> Looking for certificates that expire before 'Thu Aug 20 22:49:37 PDT 2026', but that are not expired now (Fri Nov 24 21:49:37 PST 2023).
> Error getting list of certificates
> javax.ejb.EJBException: Failed to read response
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.apply(EJBClientChannel.java:1347) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.getResult(EJBClientChannel.java:1356) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:620) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.protocol.remote.RemotingEJBClientInterceptor.handleInvocationResult(RemotingEJBClientInterceptor.java:57) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.TransactionPostDiscoveryInterceptor.handleInvocationResult(TransactionPostDiscoveryInterceptor.java:148) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.DiscoveryEJBClientInterceptor.handleInvocationResult(DiscoveryEJBClientInterceptor.java:130) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.NamingEJBClientInterceptor.handleInvocationResult(NamingEJBClientInterceptor.java:87) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.AuthenticationContextEJBClientInterceptor.call(AuthenticationContextEJBClientInterceptor.java:59) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.AuthenticationContextEJBClientInterceptor.handleInvocationResult(AuthenticationContextEJBClientInterceptor.java:52) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.TransactionInterceptor.handleInvocationResult(TransactionInterceptor.java:212) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:622) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.getResult(EJBClientInvocationContext.java:551) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBClientInvocationContext.awaitResponse(EJBClientInvocationContext.java:1003) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBInvocationHandler.invoke(EJBInvocationHandler.java:182) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.client.EJBInvocationHandler.invoke(EJBInvocationHandler.java:116) ~[jboss-client.jar:26.1.3.Final]
> 	at com.sun.proxy.$Proxy9.findCertificatesByExpireTimeWithLimit(Unknown Source) ~[?:?]
> 	at org.ejbca.ui.cli.ca.CaListExpiredCommand.getExpiredCerts(CaListExpiredCommand.java:100) [ejbca-ejb-cli.jar:EJBCA 7.9.0.2 Enterprise (working copy)]
> 	at org.ejbca.ui.cli.ca.CaListExpiredCommand.execute(CaListExpiredCommand.java:77) [ejbca-ejb-cli.jar:EJBCA 7.9.0.2 Enterprise (working copy)]
> 	at org.ejbca.ui.cli.infrastructure.command.CommandBase.execute(CommandBase.java:108) [cli-util.jar:?]
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:300) [cli-util.jar:?]
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:311) [cli-util.jar:?]
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary.findAndExecuteCommandFromParameters(CommandLibrary.java:86) [cli-util.jar:?]
> 	at org.ejbca.ui.cli.EjbcaEjbCli.main(EjbcaEjbCli.java:33) [ejbca-ejb-cli.jar:EJBCA 7.9.0.2 Enterprise (working copy)]
> Caused by: java.lang.ClassNotFoundException: com.keyfactor.util.CertificateSerializableWrapper
> 	at jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581) ~[?:?]
> 	at jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178) ~[?:?]
> 	at java.lang.ClassLoader.loadClass(ClassLoader.java:527) ~[?:?]
> 	at java.lang.Class.forName0(Native Method) ~[?:?]
> 	at java.lang.Class.forName(Class.java:398) ~[?:?]
> 	at org.jboss.marshalling.AbstractClassResolver.loadClass(AbstractClassResolver.java:129) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.AbstractClassResolver.resolveClass(AbstractClassResolver.java:110) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadClassDescriptor(RiverUnmarshaller.java:1048) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadNewObject(RiverUnmarshaller.java:1381) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:298) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:246) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadCollectionObject(RiverUnmarshaller.java:202) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.readCollectionData(RiverUnmarshaller.java:876) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:752) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.river.RiverUnmarshaller.doReadObject(RiverUnmarshaller.java:231) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.marshalling.AbstractObjectInput.readObject(AbstractObjectInput.java:41) ~[jboss-client.jar:26.1.3.Final]
> 	at org.jboss.ejb.protocol.remote.EJBClientChannel$MethodInvocation$MethodCallResultProducer.apply(EJBClientChannel.java:1323) ~[jboss-client.jar:26.1.3.Final]
> 	... 33 more
> Exception in thread "main" java.lang.NullPointerException
> 	at org.ejbca.ui.cli.ca.CaListExpiredCommand.execute(CaListExpiredCommand.java:77)
> 	at org.ejbca.ui.cli.infrastructure.command.CommandBase.execute(CommandBase.java:108)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:300)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary$Branch.execute(CommandLibrary.java:311)
> 	at org.ejbca.ui.cli.infrastructure.library.CommandLibrary.findAndExecuteCommandFromParameters(CommandLibrary.java:86)
> 	at org.ejbca.ui.cli.EjbcaEjbCli.main(EjbcaEjbCli.java:33)
>

**Comment by primetomas** [comment]
You seem to be using a CLI client from EJBCA 7.9, while the server uses 8.0.
